### PR TITLE
IDPF-87 Create boilerplate code structure

### PR DESCRIPTION
**Pull request by alsyx**
Initial code structure setup for the Identity Service.

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,289 @@
+# Python
```
Comment (marcelkornblum) [inline .dockerignore:1]: do we even need this file? docker ignore everything in gitinore by default anyway doesnt it?

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,289 @@
+# Python
```
Comment (marcelkornblum) [inline .gitignore:1]: personally im not a fan of a huge autogenerated file so much as a simple file that just includes the things we come across; it's much easier to parse visually.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,17 @@
+FROM python:3.12-bookworm
```
Comment (marcelkornblum) [inline Dockerfile:1]: i think we were going to move this file to a /docker folder but that can be a separate PR

File: config/settings/base.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    {
+        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
+    },
+]
+
+
+# Internationalization
+# https://docs.djangoproject.com/en/5.1/topics/i18n/
+
+LANGUAGE_CODE = "en-us"
```
Comment (marcelkornblum) [inline config/settings/base.py:115]: en-gb and Europe/London TZ please

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+license = "MIT"
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.12"
+django = "^5.1.1"
+django-environ = "^0.11.2"
+psycopg = "^3.2.3"
+redis = "^5.1.0"
+django-extensions = "^3.2.3"
```
Comment (marcelkornblum) [inline pyproject.toml:15]: this should be in dev dependencies

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+[tool.poetry.group.dev.dependencies]
+black = "^24.8.0"
+isort = "^5.13.2"
+pytest-django = "^4.9.0"
+
+
```
Comment (marcelkornblum) [inline pyproject.toml:22]: dont we need some config for isort, pytests, etc?

File: docs/index.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
+---
```
Comment (marcelkornblum) [inline docs/index.md:1]: do we need some config to get this generating a GH pages site?

Comment (marcelkornblum) [reply docs/index.md:1]: also a link to that in the README

File: core/tests.py [hunk 8]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
+
+# Create your tests here.
```
Comment (marcelkornblum) [inline core/tests.py:3]: can we get a simple hello world test passing in here please?

File: docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
+services:
+  web:
+    build:
```
Comment (marcelkornblum) [inline docker-compose.yml:3]: i want to add the caching stuff to this approach, as well as playwright etc, but that can be another PR

**Review by marcelkornblum — COMMENTED** [review]
Lots of small comments but overall pretty good with Vite integration etc.

File: .dockerignore (hunk 1 above)
Comment (alsyx) [reply .dockerignore:1]: I had a look, it doesn't ignore everything in `.gitignore` we need to specify what we want to ignore for docker. To start with I copied ignore files from irap. @marcelkornblum

File: Dockerfile (hunk 3 above)
Comment (0RWilliams) [reply Dockerfile:1]: Agreed with @alsyx we will do this change in a future PR, but that is the goal.

File: docs/index.md (hunk 7 above)
Comment (0RWilliams) [reply docs/index.md:1]: Agreed with @alsyx , we are gonna do this as part of [IDPF-94]( https://uktrade.atlassian.net/browse/IDPF-94)

File: docker-compose.yml (hunk 9 above)
Comment (0RWilliams) [reply docker-compose.yml:3]: Agreed with @alsyx , this will be done as part of [IDPF-105](https://uktrade.atlassian.net/browse/IDPF-105)

**Review by marcelkornblum — COMMENTED** [review]
why are we now using ASGI? Are we using it in any other projects?

**Comment by 0RWilliams** [comment]
> why are we now using ASGI? Are we using it in any other projects?

We have opted to use asgi in favour of wsgi for the performance benefits it can produce (provided we do it correctly):

Considering solutions for REST, Sam suggested django ninja to us, witch unlike DRF (django rest framework), supports asgi out of the box, and is written in a modern design, that is actively being supported with new features, rather than DRF's LTS state it's currently in.
https://www.django-rest-framework.org/community/contributing/
https://django-ninja.dev/guides/async-support/

Postgres psycopg3 supports asynchronous operations, (again  with the caveat that we do this correctly).
https://www.psycopg.org/psycopg3/docs/advanced/async.html

Django 5.1 supports async views and requests. (with a minimal penalty for use of synchronous functions)
https://docs.djangoproject.com/en/5.1/topics/async/

The hope is given this is a new project using the latest versions of Postgres, Django, and the various dependancies we can build this from the ground up to be as performant as possible given the possibly high usage the service will experience, minimising the impact this could have on consumers.

**Comment by marcelkornblum** [comment]
Since it's not boilerplate in our approach please remove it from this PR so it can be discussed on its own merits

**Review by marcelkornblum — APPROVED** [review]
🥇

**Review by 0RWilliams — APPROVED** [review]
